### PR TITLE
[FlexibleSpaceBar] Extended flexible space bar 

### PR DESCRIPTION
**Pull request by CyJaySong**
## Description
add `backgroundBuilder`, Compared to `background`, can build background widget more flexibly.
Because 'background' can only pass one fixed component and cannot display different component contents according to 'FlexibleSpaceBar' stretch, I added 'backgroundBuilder' to implement it. `Background 'and' backgroundBuilder' can't be used at the same time. Retaining 'background' ensures forward compatibility.

for issues #108655

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by CyJaySong** [comment]
@Piinks  Can you review this PR?

**Review by Piinks — COMMENTED** [review]
Hi @CyJaySong, this looks very similar to https://github.com/flutter/flutter/pull/96835, it also looks like this still does not reference an issue that this change would resolve.
Can you finish filling out the checklist in the PR description? Thanks!

**Comment by CyJaySong** [comment]
@Piinks OK, I did it. Can we merge Now?

**Comment by CyJaySong** [comment]
@kevmoo Can you review this PR?

**Comment by kevmoo** [comment]
> @kevmoo Can you review this PR?

I'm not a good person to help, sorry!

**Comment by CyJaySong** [comment]
@kevmoo can you help connect same one who can handle this PR，Piinks should be at rest now

**Comment by maheshmnj** [comment]
I was wondering instead of exposing a `backgroundBuilder`, if we could expose `onStretch` call back with Stretch details wouldn't that give more flexibility to customize FlexibleSpacebar based on the stretch?

**Comment by CyJaySong** [comment]
@maheshmnj If `onStretch` is used, it should be understood that it will only be triggered when stretching is caused by sliding; however, if `initialScrollOffset` of `ScrollController` is given a value greater than 0, `FlexibleSpaceBar` may have been stretched,`onStretch` should not be triggered

**Comment by CyJaySong** [comment]
@maheshmnj So I think `backgroundBuilder` should be used to draw components according to the actual stretch state; It has nothing to do with whether it is currently sliding

File: packages/flutter/test/material/flexible_space_bar_test.dart [hunk 1]
```diff
@@ -12,6 +12,31 @@ import 'package:flutter_test/flutter_test.dart';
...
+              },
+              collapseMode: CollapseMode.pin,
+            ),
+          ),
+        ),
+      ),
+    );
+
+    final Finder background = find.text('Background');
+    expect(background.precache(), true);
```
Comment (Piinks) [inline packages/flutter/test/material/flexible_space_bar_test.dart:36]: @CyJaySong this test does not validate or demonstrate how this utilizes the information provided by the settings.  The linked issue specifies being able to adapt to the stretch. This should be confirmed by the test.

**Review by Piinks — COMMENTED** [review]
I don't know that this is necessary. The flexible space settings are an inherited widget. If you provide a Builder to the existing `background`, you can use the context to look up the ancestor settings.

**Comment by maheshmnj** [comment]
@CyJaySong What I meant was onStretch should be called continuously while the stretch is in progress and this callback should return the scrollOffset so that user can have more flexibility to make any changes at any offset.

Heres what I was proposing

```dart
onStrecth: (offset){
  if(offset > targetOffset){
  /// change background  
  }
  if(offset > someTargetValue){
  /// do something
  }
}
```

**Comment by CyJaySong** [comment]
@Piinks I think that providing `backgroundBuilder` directly helps developers understand and use it more easily. Most users don't even know the existence of 'flexibleSpacebarSettings', and `backgroundBuilder` makes the problem easier

**Comment by CyJaySong** [comment]
@maheshmnj `onStretch` is applicable to the naming of callback functions, but here it is clear, needs to return a `widget`. For example, why don't `LayoutBuilder` use `onBuild` instead of `builder`

File: packages/flutter/lib/src/material/flexible_space_bar.dart [hunk 2]
```diff
@@ -97,6 +100,12 @@ class FlexibleSpaceBar extends StatefulWidget {
   /// Typically an [Image] widget with [Image.fit] set to [BoxFit.cover].
   final Widget? background;
 
+  /// Shown behind the [title] when expanded.
+  ///
+  /// Compared to `background`, can build background widget more flexibly.
+  final Widget Function(BuildContext context, FlexibleSpaceBarSettings settings,
```
Comment (Piinks) [inline packages/flutter/lib/src/material/flexible_space_bar.dart:106]: > Most users don't even know the existence of 'flexibleSpacebarSettings', 

If that is the case, then how should they know what to do with them here?

File: packages/flutter/lib/src/material/flexible_space_bar.dart [hunk 3]
```diff
@@ -97,6 +100,12 @@ class FlexibleSpaceBar extends StatefulWidget {
   /// Typically an [Image] widget with [Image.fit] set to [BoxFit.cover].
   final Widget? background;
 
+  /// Shown behind the [title] when expanded.
+  ///
+  /// Compared to `background`, can build background widget more flexibly.
```
Comment (Piinks) [inline packages/flutter/lib/src/material/flexible_space_bar.dart:105]: I am not sure what this documentation means. Why would the user want to use this?

**Comment by Piinks** [comment]
Please see the linked issue. There are multiple examples of how to accomplish this without adding more API. One inherits the flexible space settings, which is how they are intended to be accessed, and another is able to accomplish the test case without the settings at all just by using a LayoutBuilder. I do not think this change is necessary. It could be that we just need some documentation and example code around this.

**Comment by Hixie** [comment]
@CyJaySong Thanks for your contribution! Are you interested in continuing this work and addressing the review comments above? It's perfectly ok if you're not, we can store this PR for now and let someone else take it over when they are interested.

**Comment by CyJaySong** [comment]
I think this increase will reduce the learning cost of developers. After all, not all developers like to explore deeply hidden components such as' FlexibleSpaceBarSettings'

**Comment by CyJaySong** [comment]
@Hixie Do you think my statement is reasonable？

**Comment by Hixie** [comment]
@CyJaySong I'm not the expert here, I defer entirely to @Piinks' judgement on the issues. It sounds like @Piinks is saying that the best solution to the issue cited above is to update the documentation so as to make `FlexibleSpaceBarSettings` et al be much less "hidden" and instead be more fully documented?

**Comment by CyJaySong** [comment]
Then close it I don't mind

**Comment by Piinks** [comment]
Hey @CyJaySong, thanks for all of your contributions. Since the desired affect is already possible, do you think it would be beneficial to have better documentation or sample code? Feel free to ping me on any PRs if you would like to continue contributing, I'd be happy to continue working together. :)

**Pull request closed without merging**